### ISSUE TITLE
feat: allow to compile cairo files using only published crates

### ISSUE THREAD

**Issue by jeluard** [starkware-libs/cairo#3963]
# Feature Request

I'm trying to compile a `cairo` file into CASM using the libs published on `crates`. My attempts currently fails due to missing corelib.
Is there a way to achieve that w/o having to provide the whole corelib as files?

An example can be found here: https://github.com/jeluard/test-cairo-compiler

**Describe Preferred Solution**

Rely purely on published crates to compile cairo files.

**Comment by orizi** [comment]
If you'd use scarb as a tool it will abstract this out.

**Comment by jeluard** [comment]
Right, now my ultimate goal would be to compile `cairo` files in a webapp, so from a lib that would be compiled into wasm.
Is that something that this project would consider down the line?

**Comment by orizi** [comment]
@mkaput WDYT?

**Comment by mkaput** [comment]
@jeluard request is unfortunately not something we can solve with Scarb, because it's inherently WASM-incompatible.

The only solution to this problem I can think of is to have a `cairo-lang-corelib` crate, which [`include_dir!`s](https://docs.rs/include_dir/latest/include_dir/index.html) the corelib in compile time ([like Scarb does](https://github.com/software-mansion/scarb/blob/388db5536bf2e731be70ee64b5f251548f857ff9/scarb/src/sources/standard_lib.rs#L39-L40)). Then we need a straightforward API to feed this data into `ProjectConfig` (straight to `RootDatabaseBuilder`) as some kind of virtual directory, like @spapinistarkware did for tests some time ago.

**Comment by mkaput** [comment]
Nonetheless, is it actually possible to compile Cairo compiler to WASM? Isn't Salsa blocking this?

**Comment by jeluard** [comment]
Thanks for the info, that's helpful!

I am aware of https://github.com/cryptonerdcn/wasm-cairo/ that does compile cairo files from WebAssembly. It does rely on a fork of this project though, so unclear how many changes are required.

>  is unfortunately not something we can solve with Scarb, because it's inherently WASM-incompatible.

Out of curiosity, what is the reason for this? Is it due to the file orientation of the project, or something along those lines?

**Comment by orizi** [comment]
I think what can be done in the end - is running Sierra on WASM (including possibly compiling it to CASM) - as this is not dependent on salsa, but full Cairo would not work unless there is some implementation of salsa on WASM (which is unlikely).

salsa is the query system used by modern rust based compilers for efficient incremental building, which is good for LS support, and generally making compiler implementation simpler.

**Comment by mkaput** [comment]
> > is unfortunately not something we can solve with Scarb, because it's inherently WASM-incompatible.
> 
> Out of curiosity, what is the reason for this? Is it due to the file orientation of the project, or something along those lines?

Scarb is a CLI application (using Scarb as a Rust crate is deprecated) and is going to heavily rely on multiprocessing (though scripts and subcommands already do, we want to make `scarb build` do so, and we do not want to block ourselves from this path). Filesystem and network access could be virtualized in WASM. That would be a lot of work but still doable. But the multiprocessing and executable-binary-orientation part is the blocker.

**Comment by jeluard** [comment]
Thanks both for your comments!
